docs/SPEC.md

```diff
@@ -1364,6 +1370,83 @@ A query specification defines what data and operations to perform.
+
+Flux editions are a set of features that are enabled in Flux.
+If the edition is not enabled then the features are not enabled.
+These features would otherwise be breaking changes to Flux.
+An edition is explicitly opt-in.
+The pattern of editions allows users to migrate to new Flux versions without risk that their scripts will break.
+
+An edition is separate from a version.
+A version of Flux represents a single point in the commit history of the Flux source code.
+A user can only use a single version of Flux for a given script.
+Editions represent a set of features that are enabled for a given Flux version.
+So long as the Flux version supports all the features of the edition it can be enabled.
+A user can upgrade to a newer Flux version without being required to upgrade to the newest edition of Flux.
+
+New Flux features that require a breaking change to the syntax or semantics of Flux will always be part of a new edition.
+This means that a script can regularly update to the newest Flux version without risk of breaking because any breaking changes are explicitly opt-in.
+With editions the Flux community gets both a pattern where users can always be running the latest version of Flux and the ability to introduce new useful but otherwise breaking changes to Flux.
+
+We anticipate that there will be zero or one new editions of Flux a year. A slow cadence of new editions means users have ample time to migrate to a new edition if desired.
+Even being a few years behind on editions should only mean a few migration steps in order to have access to features introduced in the newest edition.
+
+### First Edition
+
+The first edition of Flux will be 2022.1.
+This first edition will be the only minimum required edition of Flux and represents the set of features that exist in Flux today.
+
+### Future Editions
+
+Editions will be named after the year in which they are created, with an added sequence number if more than one edition needs to be created in a single year.
+For example the first edition is `2022.1` because it is created in the year `2022` and is the first edition created in that year.
+
+### Editions are optional
+
+A Flux script may specify directly the edition it requires to function.
+Additionally the Flux runtime will allow for the edition to be specified out of band of the script, thus allowing for deployments of Flux to have control over the edition.
+
+#### Edition clause
+
+The edition clause defines the edition of Flux to be used for the current script.
+
+    EditionClause = "@" edition .
+    edition = year "." decimal_digit .
+
+It is an error for multiple files within a package specify differing editions.
+
+Example
+
+    package main
+
+    @edition 2022.2
+
+    import "array"
+
+    array.from(rows: [{_value: 1.0}])
+
+### Flux Editions and Modules
+
+Each Flux module may specify its own edition, therefore a Flux script on an earlier edition may import and consume a module that uses a newer edition of Flux.
```

Comment:
Module, package, or file? Our current syntax really only works if it's per file which is conflated with package in our current usage.

Reply:
I want it to be per module. So my thinking was the attribute must be set on the top level package for the module and if its set anywhere else its an error. Thoughts? This reduces to the most common case of per file for one file/package modules.

Reply:
There's not really a concept of a top level package since a package is just a directory and there's no root other than the folder itself.

We kind of have a module system like Rust but the structure of a Go package. I wonder if it might be worth updating the structure of what a module looks like before we have user-defined modules to be more similar to Rust because we treat it like Rust anyway.